libloki/crypto.js

```diff
@@ -324,6 +324,144 @@
     GRANT: 2,
   });
 
+  /**
+   * A wrapper around Signal's SessionCipher.
+   * This handles specific session reset logic that we need.
+   */
+  class LokiSessionCipher {
+    constructor(storage, address) {
+      this.storage = storage;
+      this.address = address;
+      this.sessionCipher = new libsignal.SessionCipher(storage, address);
+      this.TYPE = Object.freeze({
+        MESSAGE: 1,
+        PREKEY: 2,
+      });
+    }
+
+    decryptWhisperMessage(buffer, encoding) {
+      return this._decryptMessage(this.TYPE.MESSAGE, buffer, encoding);
+    }
+
+    decryptPreKeyWhisperMessage(buffer, encoding) {
+      return this._decryptMessage(this.TYPE.PREKEY, buffer, encoding);
+    }
+
+    async _decryptMessage(type, buffer, encoding) {
+      // Capture active session
+      const activeSessionBaseKey = await this._getCurrentSessionBaseKey();
+
+      if (type === this.TYPE.PREKEY && !activeSessionBaseKey) {
+        const wrapped = dcodeIO.ByteBuffer.wrap(buffer);
+        await window.libloki.storage.verifyFriendRequestAcceptPreKey(
+          this.address.getName(),
+          wrapped
+        );
+      }
+
+      const decryptFunction =
+        type === this.TYPE.PREKEY
+          ? this.sessionCipher.decryptPreKeyWhisperMessage
+          : this.sessionCipher.decryptWhisperMessage;
+      const result = await decryptFunction(buffer, encoding);
+
+      // Handle session reset
+      // This needs to be done synchronously so that the next time we decrypt a message,
+      // we have the correct session
+      try {
+        await this._handleSessionResetIfNeeded(activeSessionBaseKey);
+      } catch (e) {
+        window.log.info('Failed to handle session reset: ', e);
+      }
+
+      return result;
+    }
+
+    async _handleSessionResetIfNeeded(previousSessionBaseKey) {
+      if (!previousSessionBaseKey) {
+        return;
+      }
+
+      let conversation;
+      try {
+        conversation = await window.ConversationController.getOrCreateAndWait(
+          this.address.getName(),
+          'private'
+        );
+      } catch (e) {
+        window.log.info('Error getting conversation: ', this.address.getName());
+        return;
+      }
+
+      if (conversation.isSessionResetOngoing()) {
+        const currentSessionBaseKey = await this._getCurrentSessionBaseKey();
+        if (currentSessionBaseKey !== previousSessionBaseKey) {
```

Comment:
Need to handle the case where currentSessionBaseKey is `null`, seems like we may have some unexpected behavior on error.

Reply:
`null !== <value` will always return `true` for all non-null values. We already have a guard above to ensure that `previousSessionBaseKey` is not `null`